### ISSUE TITLE
Some functions assume ASCII/lone byte encoding while string literals seem to be utf8 encoded

### ISSUE THREAD

**Issue by fabriceleal** [urweb/urweb#145]
For instance, functions that return Basis.char (like strsub) and the declaration of uw_Basis_char itself assume strings are composed of byte-only chars. However if I hardcode a "á" character into a string literal and include it in a xml fragment, it gets displayed correctly (this probably belongs in a different issue, but I would suggest it should be rendered in html by its character entity `&aacute;` ) even though the splitting of said string using strsub yields a list with 2 "chars" with codes 195 and 161 (0xc3 0xa1).

```urweb

fun splitChs s =
    let
	val l = strlen s
    in
	if l = 0 then
	    []
	else
	    (strsub s 0) :: (splitChs (substring s 1 (l - 1)))
    end

fun joinChs cs =
    case cs of
	[] => ""
      | h :: t => (show h) ^ (joinChs t)

val lit = "á"
val l = splitChs lit
val j = joinChs l
	
fun index () =
return <xml>
  <body>
    <div>Original: <br />{[lit]}</div>
    <div>Joined back together: <br />{[j]}</div>
    <table>
      <tr>
	<th>show</th>
	<th>str1</th>
	<th>ord</th>
	<th>isalnum</th>
      </tr>

      {List.foldr (fn e acc => <xml>
	<tr>
	  <td>{[show e]}</td>
	  <td>{[str1 e]}</td>
	  <td>{[show (ord e)]}</td>
	  <td>{[show (isalnum e)]}</td>
      </tr>{acc}</xml>) <xml></xml> l}
      
    </table>
  </body>
  </xml>

```

**Comment by achlipala** [comment]
Yes, I must admit that, as a stupid American, I am guilty of adding UTF-8 support to Ur/Web (as the default) in a half-assed way.  Presumably my own stupid-American apps would be agnostic to the effects of any reasonable fixes.  Do you have particular ones to suggest?

**Comment by fabriceleal** [comment]
No fixes in particular at the moment. My particular issue is with strsub at the moment, so I could workaround by having some sort of encoding-aware-string utility library. But I think the better option would be to change the declaration of uw_Basis_char to use 4 bytes and fix the compilation errors / urweb.c functions - I could give it a shot sometime this week.

**Comment by ashalkhakov** [comment]
How about we include [utf8.h](https://github.com/sheredom/utf8.h) alongside Ur/Web C runtime? It would be a very simple matter of integrating it with the Basis.

**Comment by achlipala** [comment]
I'd be glad to include a header whose licensing terms are already compatible with Ur/Web's!

**Comment by fabriceleal** [comment]
Ok I finally started and have the bare minimum (something that makes the test app i posted above behave as intended). https://github.com/fabriceleal/urweb/commit/d957187c0c221fce524a0f73179f61f28ee36d72

I'm not sure how to implement isalnum and friends though, since that utf8.h doesnt provide such test functions.

**Comment by ashalkhakov** [comment]
There is another issue though: UTF-8 is O(n) for indexing, oh well. I think it's not such a big issue. If anybody does something with text that depends on O(1) indexing, they will probably have to use UTF-32 anyway. And that is better covered by a separate library such as ICU.

I have taken a look on the web and I don't see a readily available solution to your issue, strange. What you could do for now is just to check if the incoming codepoint is small enough so that you could fall back to ASCII and report "don't know" or some such for bigger codepoints.

There's also [this](https://daniel.haxx.se/blog/2018/01/30/isalnum-is-not-my-friend/) -- `isalnum` and friends confined to ASCII.

**Comment by fabriceleal** [comment]
> UTF-8 is O(n) for indexing
Yeah that's itching me as well. Not sure how @achlipala feels about it.  Using UTF-8 seems like a nice compromise since urweb is already using uw_Basis_string as the type used for request headers, for instance (point being, migrating to a fixed-width encoding might a lot of work). Or maybe we could keep uw_Basis_string as ASCII only and use it for the most "basic" stuff and create a uw_Basis_utfstring type. But in this case I think it could be worthwhile to use UTF-32. 

About isalnum and friends: I took a look at mono and it seems the way this is implemented is by having a big table mapping codepoints to categories https://github.com/mono/mono/blob/d065dc9292e76af34fe2bce3ab2d629acbe1e38c/mcs/class/referencesource/mscorlib/system/globalization/charunicodeinfo.cs 

https://github.com/mono/mono/blob/master/mcs/class/corlib/resources/charinfo.nlp

**Comment by ashalkhakov** [comment]
@fabriceleal I've checked the way it's handled in Mono and ... ugh. Those are huge tables. What do you think if we just use ICU instead of `utf8.h` instead? It does contain char predicates: http://icu-project.org/apiref/icu4c/uchar_8h.html

>  Using UTF-8 seems like a nice compromise

My thoughts exactly. And the O(n) issue mostly depends on what kind of processing you do.

**Comment by fabriceleal** [comment]
Yeah I will try to give ICU a try this weekend, and report back

**Comment by fabriceleal** [comment]
Ok, I have the bare minimum, now with libicu: https://github.com/fabriceleal/urweb/commit/fb70fc5b8dfc2694fc6d168e9c5724a975665ffd

I installed libicu-dev and manually inserted the necessary libs into liburweb_la_LIBADD in the src/c/Makefile (-licui18n -licuuc -licudata), since I don't know any better! :stuck_out_tongue: 

isalnum seems to work at least for the character "à", however I noticed also an issue with trying to use `#"à"` in urweb code, where the compiler complains "Character constant is zero or multiple characters" - which I suppose is due to the parser itself not expecting a multi byte char.

**Comment by achlipala** [comment]
Thanks for the changes.  I see a few spans of comments, indicating intermediate markers in your migration process, where probably those spans of code can just be deleted.  Can you please change it accordingly?

As for changing the build process, the right fix is editing `Makefile.am` in any directory where you are tempted to edit `Makefile`, then rerunning `autogen,sh` at the top level of the repo directory structure.  Could you please make that change so that, e.g., everything builds in Travis CI?  It probably requires editing `.travis.yml` to install the new dependency package.

The support for UTF-8 literals in Ur/Web code is a bit of a pain in the ass!  Does anyone out there who is used to using UTF-8 also have the requisite SML expertise?  Alternatively, we could merge this change without yet supporting literals in Ur/Web code.  (Actually, I wonder what else might need to change in the compiler proper, for fully reasonable UTF-8 handling.)

Once the changeset has addressed my requests, I'll invest a bit in reading documentation for the library you've pulled in, so I can check that the code seems to be doing the right thing.  Filing a PR at that point sounds like a good idea.

**Comment by ashalkhakov** [comment]
On a first glance it seems like line 257 of `urweb.lex` is the culprit here. If the lexer would only preserve the 8-bit sequences of input, carrying those into output as-is, then I guess UTF-8 would work just fine (source file in UTF-8 -> lexer that preserves the octets -> C file in UTF-8/JS file in UTF-8 -> should work fine at runtime).

**Comment by fabriceleal** [comment]
Ok so I migrated the string manipulation functions in urweb.c (checkpoint: https://github.com/fabriceleal/urweb/commit/e11155b82ad16029a3257cdfa9bfc627fec49ef4) and I think I got them right: I think I should write some unit tests before I submit a PR? 

Also some functions (uw_Basis_atom, uw_Basis_css_url, uw_Basis_property) look like they only want "ASCII strings", so I'm testing if the char is a single-byte-utf8-char (U8_IS_SINGLE) - should I keep this? take it off? Should I assume testing isalnum or a subset of it means "ASCII only" ?

I did some light testing with some japanese characters and portuguese characters and it _seems_ ok. The jsify and htmlify could require migration, I have to take a careful look at them; sqlify seems to work without issues on my postgres database (values are stored and retrieved correctly), didn't test sqlite though.

I'm using macros / functions from here http://icu-project.org/apiref/icu4c/utf8_8h.html and here http://icu-project.org/apiref/icu4c/uchar_8h.html ; this was the only documentation I used. Most (almost all?) of the rest of the library seems to be dedicated to fixed length encodings

**Comment by ashalkhakov** [comment]
That's some good work there, @fabriceleal !

On the topic of testing, you should be able to at least run some automated tests (if you install python, selenium, and Chrome). These tests will run with SQLite and will exercise the server code of urweb. Ping me if you have any issues with the setup. There are many more tests out there but they will require manual inspection to see if things still work.

**Comment by fabriceleal** [comment]
Ok made some unit tests. Some tests fail on my end, but they're the same ones that fail on master. I rebased everything into a single commit and created the PR https://github.com/urweb/urweb/pull/146
